### ISSUE TITLE
Optional arguments are required when using ParamSpec

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1285]
**Describe the bug**
Callables annotated with `ParamSpec` erroneously expect optional arguments to be provided.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `qux` would not be required.

**Screenshots or Code**

```python
from typing import Callable, TypeVar
from typing_extensions import ParamSpec

V = TypeVar('V')
P = ParamSpec('P')

def foo(fn: Callable[P, V]) -> Callable[P, V]:
    ...

def bar(baz: str, qux: str = '') -> str:
    ...

# 'Argument missing for parameter "qux"'
foo(bar)('')

C = TypeVar('C', bound=Callable[..., object])

def quux(fn: C) -> C:
    ...

# Ok
quux(bar)('')

```

**VS Code extension or command-line**
Either, 1.1.95

**Comment by erictraut** [comment]
@gvanrossum, I'm interested in your thoughts on this one. [PEP 612](https://www.python.org/dev/peps/pep-0612/) doesn't appear to say anything about parameters with default values. How are they meant to be handled in a collection of parameters bound to a ParamSpec?

**Comment by erictraut** [comment]
Adding this support is straightforward. Here's a [PR](https://github.com/microsoft/pyright/pull/1286). I'll give Guido a chance to confirm that this is the intended behavior before I merge it.

**Comment by gvanrossum** [comment]
The intention as I always understood it was that `foo` should take a function with an *arbitrary* signature and return one with exactly the same signature (i.e. `foo` should behave the same as `quux`).

While a regular `Callable` only supports mandatory positional arguments (with the exception of `Callable[..., RT]`), `ParamSpec` is supposed to capture arbitrary signatures. The closest PEP 612 comes to hinting at this is I think the first example of the [`semantics`](https://www.python.org/dev/peps/pep-0612/#id14) section, which makes it clear that argument *names* are preserved. It wouldn't make sense to preserve argument names but not default values -- or at least the presence of defaults.

I wouldn't be opposed to asking @mrkmndz if he's okay with us adding a section to the PEP that clarifies this. Just an example somewhere would be sufficient -- this PEP mainly seems to use examples to describe semantics. I tried to see what Pyre does, but I gave up -- maybe it's still called `ParameterSpecification` there, as in the first draft?

**Comment by erictraut** [comment]
OK, thanks for the clarification. That justification makes sense to me. I'll merge the PR.

@layday, thanks for the suggestion. This change will be in the next release of pyright and pylance.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.96, which I just published. It will also be in the next published release of Pylance.

**Comment by layday** [comment]
Thanks for addressing this issue.  However, this still produces an error for me:

```py
from typing import Callable, TypeVar
from typing_extensions import ParamSpec

T = TypeVar("T")
P = ParamSpec("P")
C = TypeVar("C", bound=Callable[..., object])

def foo(f: Callable[P, T]) -> Callable[P, T]:
    ...

def bar(f: C) -> C:
    ...


from tempfile import NamedTemporaryFile

NamedTemporaryFile(delete=False)

# Argument missing for parameter "mode"
foo(NamedTemporaryFile)(delete=False)

# Ok
bar(NamedTemporaryFile)(delete=False)

```

I assume it has something to do with the interaction with `overload`.

**Comment by mrkmndz** [comment]
@gvanrossum @erictraut Yes, you're right, PEP 612 did intend to preserve default/optional arguments, and I'd be more than happy for any of you to modify the PEP to make that clear!

**Comment by erictraut** [comment]
@layday, your example above is another case that apparently is not contemplated by PEP 612. It involves an overloaded function where signatures differ between the various overloads. It's not clear what a ParamSpec should capture in this case. Pyright's current implementation captures the signature of the first overload. That's probably not correct, but I'm not sure it's any more or less correct than other options.

@mrkmndz, what behavior did you have in mind for overloaded functions that are captured by a ParamSpec? Perhaps it should be considered an error condition?

**Comment by layday** [comment]
Considering that this is the exact use case the PEP opens with, rewrapping the return type of a function as I am trying to do, if we cannot use `ParamSpec` to rewrap the return type of all function variants, that would be rather unfortunate - overloads are everywhere in the stdlib.

**Comment by gvanrossum** [comment]
Agree with @layday. Assuming there's an internal way to spell an overloaded function, `ParamSpec` should capture that. This is why it's a magic cookie that can only be used in specific cases -- but the limitations deal with where `P` can occur in types, not with the signatures it catches.

**Comment by erictraut** [comment]
The problem is that a ParamSpec doesn't capture the full signature of a function. It captures only the input parameters. Another independent type variable (`T` in the example above) captures the return type.

Consider the example provided above. `P` and `T` are independent type variables, and we're matching against the type expression `Callable[P, T]`. Let's assume that the overloaded function has different return types for each overload (which is typically the case). If so, what type would `T` be bound to in this example?

I don't see a way to use ParamSpec with overloads. I think our options are 1) say that it's illegal and emit an error if it's attempted, or 2) specify a deterministic rule like "use the first overload" for matching.

**Comment by gvanrossum** [comment]
Another option (not saying it's the best) would be to say `T` becomes the union of the return types of the individual overloads. But this still doesn't really solve the problem. Let's create an example:
```
@overload
def cat(a: str, b: str) -> str: ...
@overload
def cat(a: bytes, b: bytes) -> bytes: ...

**Comment by erictraut** [comment]
I'll point out that if the intent is to simply copy the entire signature of the callable argument as is done with `foo` above, a `ParamSpec` isn't required or recommended. Instead, `foo` could be defined as:

```python
T = TypeVar("T", bound=Callable[..., Any])
def foo(f: T) -> T: ...
```

`ParamSpec` is needed only in cases where 1) you need to separate the input parameters from the return type, or 2) you need to add or remove input parameters (a'la `Concatenate`). Neither of those operations make sense on an overloaded function.

I agree that declaring this case as an error is better than picking the first overload. I'll create a new bug to track this.

**Comment by layday** [comment]
My specific intent is to take a synchronous, overloaded function and return a couroutine function:

```py
from __future__ import annotations

from collections.abc import Awaitable
from tempfile import NamedTemporaryFile
from typing import Callable, TypeVar

T = TypeVar('T')

def foo(f: Callable[..., T]) -> Callable[..., Awaitable[T]]:
    ...

# Here, this is IO[str], the first overload variant's return type; it should be IO[bytes]
reveal_type(foo(NamedTemporaryFile)())
```

This cannot be accomplished neither with a bounded `Callable` (type variables cannot be nested) nor with `ParamSpec` (overloads are discarded).

**Comment by erictraut** [comment]
@mrkmndz, if I understand you correctly, you're not treating `P` and `T` as independent type variables; you're treating them as somehow "entangled" (in the quantum sense of the term).

**Comment by mrkmndz** [comment]
Yup @erictraut , that's true, but this is not restricted to `ParamSpec`s.  In general Pyre handles constraint solving by generating disjunctions of conjunctions (like {(P=(a: str, b: str) AND T = str) OR (P=(a:bytes, b:bytes) AND T=bytes)}.).

This comes up with solving overloaded functions against (non `ParamSpec`) generics, as well as solving against unions that contain type variables.

Here's a contrived example, but this stuff did in fact come up in practice in more complex forms.
```
def foo(x: Union[Tuple[List[T1], T2], Tuple[T2, List[T1]]]) -> Tuple[T1, T2]: ...

x: Tuple[List[int], List[str]]
foo(x) -> {(T1 = int AND T2=List[str]) OR (T1 = List[int] AND T2 = str)}
```

**Comment by erictraut** [comment]
@mrkmndz, that's an interesting approach. Does it work with `Concatenate` as well? Is the "concatenate" operator applied to each of the overloads independently?

FYI, Pyright handles ambiguous cases like the one in your example by "scoring" the valid constraint sets and picking the simplest of them. In the event of a tied score (which I've never seen happen in the wild but does occur in the example you provided above), it picks the first one it finds. So in your example, Pyright evaluates the type of `foo(x)` as `Tuple[str, List[int]]`, although `Tuple[List[str], int]` would be correct too. It sounds like Pyre doesn't do any similar scoring but instead generates all possible solutions even if some are more complex than others. And then it carries the first such solution forward. Is that correct?

Interestingly, mypy struggles with your example above. It emits an (incorrect) error indicating that argument 1  of "foo" is incompatible, and it evaluates the type of `foo(x)` as `Tuple[<nothing>, <nothing>]`.
